signature/src/mechanism/cosign/mod.rs

```diff
@@ -0,0 +1,380 @@
+    /// If it is neither of them, return `Error`. Otherwise, return `Ok()`
+    fn check_reference_rule_types(&self) -> Result<()> {
+        match &self.signed_identity {
+            Some(rule) => match rule {
+                PolicyReqMatchType::MatchRepository
+                | PolicyReqMatchType::ExactRepository { .. } => Ok(()),
+                p => Err(anyhow!("Denied by {:?}", p)),
+            },
+            None => Ok(()),
+        }
+    }
+
+    /// Verify the cosign-signed image. There will be three steps:
+    /// * Get the pub key.
+    /// * Download the cosign-signed image's manifest and its digest. Calculate its
+    /// signature's image.
+    /// * Download the signature image, gather the signatures and verify them
+    /// using the pubkey.
+    /// If succeeds, the payloads of the signature will be returned.
+    async fn verify_signature_and_get_payload(&self, image: &mut Image) -> Result<Vec<SigPayload>> {
+        // Get the pubkey
+        let key = match (&self.key_data, &self.key_path) {
+            (None, None) => return Err(anyhow!("Neither keyPath nor keyData is specified.")),
+            (None, Some(key_path)) => read_key_from_url(key_path).await?,
+            (Some(key_data), None) => key_data.as_bytes().to_vec(),
+            (Some(_), Some(_)) => return Err(anyhow!("Both keyPath and keyData are specified.")),
+        };
+
+        let mut client = sigstore_rs::cosign::ClientBuilder::default().build()?;
+
+        let auth = &sigstore_rs::registry::Auth::Anonymous;
+        let image_ref = image.reference.whole();
+
+        // Get the cosign signature "image"'s uri and the signed image's digest
+        let (cosign_image, source_image_digest) = client.triangulate(&image_ref, auth).await?;
+
+        // Get the signature layers in cosign signature "image"'s manifest
+        let signature_layers = client
+            .trusted_signature_layers(auth, &source_image_digest, &cosign_image)
+            .await?;
+
+        // By default, the hashing algorithm is SHA256
+        let pub_key_verifier = PublicKeyVerifier::new(&key, SignatureDigestAlgorithm::Sha256)?;
+
+        let verification_constraints: VerificationConstraintVec = vec![Box::new(pub_key_verifier)];
+
+        let res = verify_constraints(&signature_layers, verification_constraints.iter());
+
+        match res {
+            Ok(()) => {
+                // gather the payloads
+                let payloads = signature_layers
+                    .iter()
+                    .map(|layer| SigPayload::from(layer.simple_signing.clone()))
+                    .collect();
+                Ok(payloads)
+            }
+            Err(SigstoreVerifyConstraintsError {
+                unsatisfied_constraints,
+            }) => Err(anyhow!("{:?}", unsatisfied_constraints)),
+        }
+    }
+}
+
+async fn read_key_from_url(url: &str) -> Result<Vec<u8>> {
+    // TODO: Do we need define a new URL scheme
+    // named `kbs://` to indicate that the key
```

Comment:
Yes, thanks for the link. I changed the function name to avoid confusing. We may take some effort to support more source of the key, like different urls, or a directly path to local file, etc.